theano/tensor/opt.py

```diff
@@ -3902,6 +4015,110 @@ def local_elemwise_sub_zeros(node):
+    # Elemwise[minimum](X.shape[i], 0) -> 0
+    # Elemwise[minimum](0, X.shape[i]) -> 0
+
+    # The shape can be replaced with sum of shapes
+    Elemwise[LT](sum([anything that is shapes]), 0) -> Elemwise[zeros](X)
+    Elemwise[GE](sum([anything that is shapes]), 0) -> Elemwise[ones](X)
+
+    """
+    if not isinstance(node.op, T.Elemwise):
+        return
+    if node.op.scalar_op.nin != 2:
+        return
+    # Elemwise[{LT,GT}](X, X) -> Elemwise[zeros](X)
+    if (isinstance(node.op.scalar_op, (scalar.LT, scalar.GT)) and
+        node.inputs[0] is node.inputs[1]):
+        return [T.zeros_like(node.outputs[0], dtype=node.outputs[0].type.dtype)]
+    # Elemwise[{LE,GE}](X, X) -> Elemwise[ones](X)
+    if (isinstance(node.op.scalar_op, (scalar.LE, scalar.GE)) and
+        node.inputs[0] is node.inputs[1]):
+        return [T.ones_like(node.inputs[0], dtype=node.outputs[0].type.dtype)]
+    # Elemwise[{minimum,maximum}](X, X) -> X
+    if (isinstance(node.op.scalar_op, (scalar.Minimum, scalar.Maximum)) and
+        node.inputs[0] is node.inputs[1]):
+        return [node.inputs[0]]
+
+    # Elemwise[LT](X.shape[i], 0) -> Elemwise[zeros](X)
+    if (isinstance(node.op.scalar_op, scalar.LT) and
+        node.inputs[0].owner and
+        isinstance(node.inputs[0].owner.op, Shape_i) and
+        T.extract_constant(node.inputs[1]) == 0):
+        return [T.zeros_like(node.outputs[0])]
+    # Elemwise[GE](X.shape[i], 0) -> Elemwise[ones](X)
+    if (isinstance(node.op.scalar_op, scalar.GE) and
+        node.inputs[0].owner and
+        isinstance(node.inputs[0].owner.op, Shape_i) and
+        T.extract_constant(node.inputs[1]) == 0):
+        return [T.ones_like(node.outputs[0])]
+    # Elemwise[maximum](X.shape[i], 0) -> X.shape[i]
+    if (isinstance(node.op.scalar_op, scalar.Maximum) and
+        node.inputs[0].owner and
+        isinstance(node.inputs[0].owner.op, Shape_i) and
+        T.extract_constant(node.inputs[1]) == 0):
+        return [node.inputs[0]]
+    # Elemwise[maximum](0, X.shape[i]) -> X.shape[i]
+    if (isinstance(node.op.scalar_op, scalar.Maximum) and
+        T.extract_constant(node.inputs[0]) == 0 and
+        node.inputs[1].owner and
+        isinstance(node.inputs[1].owner.op, Shape_i)):
+        return [node.inputs[1]]
+    # Elemwise[minimum](X.shape[i], 0) -> 0
+    if (isinstance(node.op.scalar_op, scalar.Minimum) and
+        node.inputs[0].owner and
+        isinstance(node.inputs[0].owner.op, Shape_i) and
+        T.extract_constant(node.inputs[1]) == 0):
+        return [T.zeros_like(node.outputs[0])]
+    # Elemwise[minimum](0, X.shape[i]) -> 0
+    if (isinstance(node.op.scalar_op, scalar.Minimum) and
+        T.extract_constant(node.inputs[0]) == 0 and
+        node.inputs[1].owner and
+        isinstance(node.inputs[1].owner.op, Shape_i)):
+        return [T.zeros_like(node.outputs[0])]
+
+    # Elemwise[LT](sum([anything that is shapes]), 0) -> Elemwise[zeros](X)
```

Comment:
rename sum to add. sum make me thing this is the sum reduction, but it isn't this case.